### DISCUSSION TITLE
Support for Agents with Tool Calls in Langfuse Prompt Management UI

### DISCUSSION THREAD

**Discussion by laisalvescw** [langfuse/langfuse#4674] [Ideas]
### Describe the feature or potential improvement

Hello Langfuse Team and Community, I would like to propose an enhancement to the Langfuse UI for prompt management that would greatly benefit developers working with agents that utilize tool calls.

Currently, the Langfuse prompt management UI provides a robust interface for managing prompts. However, as far as I can tell, it does not support agents that rely on tool calls. This limitation poses a challenge for developers like myself who have integrated tool-dependent agents into our workflows. Also, while we can manage prompts through the Langfuse UI, there isn't a straightforward way to connect our custom agent APIs to the prompt management system.

Here are two potential solutions to address these limitations:

**Solution 1: Template-Based Tool Configuration**

Introduce a new section in the Langfuse UI where developers can define tool configurations as templates.

Allow users to create "tool profiles" that specify:
- Tool name and description
- Required parameters and their types
- Expected response format
- Authentication requirements

The prompt management interface would then allow users to reference these tool templates within their prompts.
When testing prompts, Langfuse would simulate tool calls using mock data or allow connection to actual tool endpoints.
This would maintain the current UI structure while adding support for tool-dependent agents.

**Solution 2: Agent API Bridge**
Create a new integration layer that acts as a bridge between Langfuse and custom agent APIs.

Developers would provide:
- Agent API endpoint
- Authentication details
- Tool specifications (as a schema)
- Input/output format requirements

The Langfuse UI would then:
- Auto-detect available tools from the agent API
- Allow direct testing against the actual agent implementation
- Provide real-time feedback on tool call success/failure
- Track and analyze tool usage patterns

This solution would be more comprehensive but would require more significant changes to the Langfuse architecture.

Both solutions address the core issues but take different approaches:
- Solution 1 focuses on managing tool configurations within Langfuse.
- Solution 2 emphasizes direct integration with existing agent implementations.

The choice between them would depend on factors like:
- Development resources available
- Desired level of integration
- Security requirements
- Performance considerations

The second option works better for my case.

Thank you for considering this enhancement.
Best regards.

**Comment by marcklingen** [comment]
Thanks for sharing! The tool-library that can be referenced in prompts within prompt management seems like a very good extension of prompt management. See this thread for additional context: https://github.com/orgs/langfuse/discussions/2624

The second aspect seems interesting as well but would require a standard implementation of tool discovery. Do you have a standard in mind?

**Reply by laisalvescw** [reply]
Sure! I can share with you on our private channel on slack.

**Comment by sfarthin** [comment]
An important aspect of tool calling is ensuring that the downstream application aligns with the function signature of the tools. There are changes that can break this API and others that do not. Enhancing function names and tool descriptions is a form of improvement that I believe would be most beneficial for Langfuse, as these changes typically do not affect the API compatibility. However, changes to input/output configurations require meticulous coordination with the software application.

To maintain stability in production, I suggest that any "tool configuration" should lock the schema, ensuring that Langfuse users do not inadvertently promote a "tool template" with a different signature. In the case of necessary changes to input/output configurations, users should create a completely new "tool set" with a separate identifier, requiring careful consideration from developers to implement these changes.

Additionally, by identifying functions with IDs at the application level, Langfuse could allow for adjustments to the function names without affecting the function signature. This flexibility would enable subtle improvements while safeguarding the integrity of production systems.

**Reply by marcklingen** [reply]
This is great, 100% agree with your take on signature/config on the tool template level vs descriptions that help improve tools without breaking logic that depends on them.
